core/src/main/scala/org/apache/spark/shuffle/ShuffleWriter.scala

```diff
@@ -17,18 +17,466 @@
+      // recommended to avoid any blocking operations in the eventloop;
+      // 2. The actual connection creation is a task that gets added to the task queue of another
+      // eventloop which could have eventloops eventually blocking each other.
+      // Once the blockPushListener is notified of the block push success or failure, we
+      // just delegate it to block-push-threads.
+      def handleResult(result: PushResult): Unit = {
+        submitTask(() => {
+          if (updateStateAndCheckIfPushMore(
+            sizeMap(result.blockId), address, remainingBlocks, result)) {
+            pushUpToMax(SparkEnv.get.blockManager.blockStoreClient)
+          }
+        })
+      }
+
+      override def onBlockFetchSuccess(blockId: String, data: ManagedBuffer): Unit = {
+        logTrace(s"Push for block $blockId to $address successful.")
+        handleResult(PushResult(blockId, null))
+      }
+
+      override def onBlockFetchFailure(blockId: String, exception: Throwable): Unit = {
+        // check the message or it's cause to see it needs to be logged.
+        if (!errorHandler.shouldLogError(exception)) {
+          logTrace(s"Pushing block $blockId to $address failed.", exception)
+        } else {
+          logWarning(s"Pushing block $blockId to $address failed.", exception)
+        }
+        handleResult(PushResult(blockId, exception))
+      }
+    }
+    shuffleClient.pushBlocks(address.host, address.port, blockIds.toArray,
+      sliceReqBufferIntoBlockBuffers(request.reqBuffer, request.blocks.map(_._2)),
+      blockPushListener)
+  }
+
+  /**
+   * Given the ManagedBuffer representing all the continuous blocks inside the shuffle data file
+   * for a PushRequest and an array of individual block sizes, load the buffer from disk into
+   * memory and slice it into multiple smaller buffers representing each block.
+   *
+   * With nio ByteBuffer, the individual block buffers share data with the initial in memory
+   * buffer loaded from disk. Thus only one copy of the block data is kept in memory.
+   * @param reqBuffer A {{FileSegmentManagedBuffer}} representing all the continuous blocks in
+   *                  the shuffle data file for a PushRequest
+   * @param blockSizes Array of block sizes
+   * @return Array of in memory buffer for each individual block
+   */
+  private def sliceReqBufferIntoBlockBuffers(
+      reqBuffer: ManagedBuffer,
+      blockSizes: Seq[Long]): Array[ManagedBuffer] = {
+    if (blockSizes.size == 1) {
+      Seq(reqBuffer).toArray
+    } else {
+      val inMemoryBuffer = reqBuffer.nioByteBuffer()
+      val blockOffsets = new Array[Long](blockSizes.size)
+      var offset = 0L
+      for (index <- blockSizes.indices) {
+        blockOffsets(index) = offset
+        offset += blockSizes(index)
+      }
+      blockOffsets.zip(blockSizes).map {
+        case (offset, size) =>
+          new NioManagedBuffer(inMemoryBuffer.duplicate()
```

Comment:
Would it be better to use `FileSegmentManagedBuffer` here in terms of memory usage?

Reply:
Here, we are loading the shuffle block data generated by the mappers on disk into memory.
There are a few considerations being made here:
1) The current push shuffle RPC is for pushing individual shuffle blocks. This simplifies the retry and error handling logic on the server side. On the client side, we could have skipped loading shuffle blocks into memory, and just directly transmit individual shuffle blocks from disk to the server side. However, by doing this, we would again be subject to the issue of small shuffle block size hurting disk I/O throughput (especially HDD). By batch-loading a group of blocks in the push request with a single I/O, we can control the size of data read from disk as the block batch size. Once loaded in memory (line 268), we further slice the buffer into multiple ones each corresponding to a single block.
2) In terms of the memory usage, it's actually bounded. In `updateStateAndCheckIfPushMore` at line 327, you can see that we actually wait for an entire batch to be fully pushed (remainingBlocks.isEmpty) before processing the next batch. This way, the memory usage is bounded to # fetch threads * block batch size.